### DISCUSSION TITLE
ESPNow received MAC address error

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#9860] [General]
Curious issue using ESPNow; the irecv() function returns the MAC of the sender, but with the first byte (MSB) incremented by 2 from the original.
In my case, the first byte of the sender is "c4", but it's received as "c6"; all the other bytes are received correctly.

I seem to recall posting about this issue some time ago, and I believe that Glenn20 resolved the issue. Unfortunately, the resolved issue was in a PM to Glenn on the original micropython forum, and the PM's are no longer available.

Has anyone experienced this problem, and knows the solution?

**Comment by davefes** [comment]
I recall that a off-by-two is the difference between AP and STA modes.
davef

**Reply by mmliam** [reply]
Thanks dave, but I was about to make an edit, to make it clearer. The erroneous byte is the Most Significant Byte (MSB); the STA & AP generally differ by one in the LSB.

**Reply by mmliam** [reply]
Hey dave, The issue is solved; turns out we were both right. The STA & AP of the ESP32 affect the LSB by one digit, The STA & AP of the ESP8266 affect the MSB by two digits. Talk about lack of consistency.

Here's where it really gets weird; for the ESP-01S, irecv returns the STA mac. For the ESP-07, irecv returns the AP mac. These are both ESP8266 based modules.

**Comment by davefes** [comment]
Yet another trap ...  

For your last statement  are both units set to either STA or AP?

**Reply by mmliam** [reply]
Both were STA.

**Reply by mmliam** [reply]
Hey dave, I found the problem; now irecv() for both the ESP-01s & ESP-07 returns the STA mac. I was activating AP to set the channel and the STA and then I would deactivate the AP. Apparently I was doing testing on the ESP-07 code some time ago and commented out the AP deactivation; and then promptly forgot it was active. When the deactivation was uncommented, irecv() received the STA mac. It was bothering me that two ESP8266 based devices would function differently.

BTW: I just discovered a great free code editor, that can be customized for various languages it's called Notepad++. If you haven't heard of it here's a link: https://notepad-plus-plus.org/downloads/v8.4.6/
What's really useful is when you make it the default editor for .py, every time you click on a different code file it opens it an a new tab. You can quickly compare code by alternately clicking on the respective tabs.

**Comment by davefes** [comment]
Great there was a reason for the difference!

I used NotePad++ years ago before I shifted to Linux.  Does it actually do a visible diff on the files?  I use Meld.

**Reply by mmliam** [reply]
You select your language and it formats the text. Indention's are even displayed with a faint line emphasizing the grouping. You can even define you own language formatting.

Give it a try again, it's free.

**Reply by davefes** [reply]
Meld does that as well.  What it didn't pick up was a reboot() with correct indentation but at the wrong level in the code :)

**Reply by mmliam** [reply]
Discovered something interesting with uPy on the ESP8266 based module. I removed all references to the AP object. Apparently you don't need to fix a channel using an AP object; the STA object will find the router channel by itself. I've noticed this before, but wanted to recheck.

**Reply by davefes** [reply]
Are you using wifi.py in 
https://github.com/glenn20/micropython-espnow-utils

**Reply by mmliam** [reply]
I've seen those utilities, but I'm not using them. I shouldn't be surprised that the STA object finds the router channel; because it does scan for it. The advantage of setting the channel using the AP object is eliminating the scan time; and that does make a significant difference.

**Reply by davefes** [reply]
The last line I have read somewhere else, maybe on the old forum.

I have been using wifi.py on all my ESPNow boards since it came out ... with no problems.

**Reply by mmliam** [reply]
As far as eliminating the scan time; I determined from the first use of the ESP devices when I was still using wifi connections. After I switched to ESPNow it became less of an issue, but still significant for one-off activation's. I'm just annoyed that the uPy for the ESP8266 doesn't allow channel selection for the STA as it does for the ESP32. Frankly, after reviewing the source for wifi.py it appears to have more overhead than I would like. My code is lean-n-mean and has been functioning perfectly. I just need to remember when I comment-out code for experiments - lol.

Right now my goal is to completely eliminate some legacy software & hardware. To that end I have devices running ESPNow, communicating with a Gateway running ESPNow & MQTT, that's interfaced over wifi to Node-Red running on a laptop.  With Node-Red I've duplicated many of the legacy software features while simultaneously interacting with the remaining features on legacy software. As I incorporate more of the features into Node-Red I eliminate them from the legacy software. In other words I can gradually transition from legacy to Node-Red. Then I'll gradually replace the remaining legacy hardware will my own.

**Reply by glenn20** [reply]
> I'm just annoyed that the uPy for the ESP8266 doesn't allow channel selection for the STA as it does for the ESP32.

Yeah - I'm annoyed by that too. It is just not supported in the esp8266 SDK (and also wasn't supported on the esp32 IDF till some time in version 4+). For wifi it doesn't make sense to set the channel on the STA. Espressif added support for this in the esp32 IDF specifically for users of ESPNOW.

You can set the channel on the ESP8266 STA by activating the STA, then the AP, setting the channel of the AP (which will also set the STA channel) and then deactivate the AP. The STA shoud continue to operate on the channel you have selected (unless of course you connect the STA to a wifi access point :-). But - that is annoying.
